examples/ckks/ckks_tutorial/main.go

```diff
@@ -0,0 +1,787 @@
+
+	// To store and manage the loading of evaluation keys, we instantiate a struct that complies to the `rlwe.EvaluationKeySetInterface` Interface.
+	// The package `rlwe` provides a simple struct that complies to this interface, but a user can design its own struct compliant to the `rlwe.EvaluationKeySetInterface`
+	// for example to manage the loading/saving/persistence of the keys in the memory.
+	evk := rlwe.NewMemEvaluationKeySet(rlk)
+
+	// ====================
+	// Plaintext Generation
+	// ====================
+	//
+	// We use the default number of slots, which is N/2.
+	// It is possible to use less slots, however it most situations, there is no reason to do so.
+	LogSlots := params.LogMaxSlots()
+	Slots := 1 << LogSlots
+
+	// We generate a vector of `[]complex128` with both the real and imaginary part uniformly distributed in [-1, 1]
+	/* #nosec G404 -- this is a plaintext vector  */
+	r := rand.New(rand.NewSource(0))
+	values1 := make([]complex128, Slots)
+	for i := 0; i < Slots; i++ {
+		values1[i] = complex(2*r.Float64()-1, 2*r.Float64()-1)
+	}
+
+	// We allocate a new plaintext, at the maximum level.
+	// We can allocate plaintexts at lower levels to optimize memory consumption for operations that we know will happen at a lower level.
+	// Plaintexts (and ciphertexts) are by default created with the following metadata:
+	//   - `Scale`: `params.DefaultScale()` (which is 2^{45} in this example)
+	//   - `EncodingDomain`: `rlwe.SlotsDomain` (this is the default value)
+	//   - `LogSlots`: `params.MaxLogSlots` (which is LogN-1=13 in this example)
+	// We can check that the plaintext was created at the maximum level with pt1.Level().
+	pt1 := ckks.NewPlaintext(params, params.MaxLevel())
+
+	// Then we need to instantiate the encoder, which will enable us to embed our `values` of type `[]complex128` on a `rlwe.Plaintext`.
+	// By default the encoder will use the params.DefaultPrecision(), but a user can specify a custom precision as an optional argument,
+	// for example `ckks.NewEncoder(params, 256)`.
+	ecd := ckks.NewEncoder(params)
+
+	// And we encode our `values` on the plaintext.
+	// Note that the encoder will check the metadata of the plaintext and adapt the encoding accordingly.
+	// For example, one can modify the `Scale`, `EncodingDomain` or `LogSlots` fields change the way the encoding behaves.
+	if err = ecd.Encode(values1, pt1); err != nil {
+		panic(err)
+	}
+
+	// =====================
+	// Ciphertext Generation
+	// =====================
+	//
+	// To generate ciphertexts we need an encryptor.
+	// An encryptor will accept both a secret key or a public key,
+	// in this example we will use the public key.
+	enc := ckks.NewEncryptor(params, pk)
+
+	// And we create the ciphertext.
+	// Note that the metadata of the plaintext will be copied on the resulting ciphertext.
+	ct1, err := enc.EncryptNew(pt1)
+	if err != nil {
+		panic(err)
+	}
+
+	// It is also possible to first allocate the ciphertext the same way it was done
```

Comment:
Maybe specify what differs between the two methods of allocation

Reply:
Done